internal/user/model.go

```diff
@@ -1,52 +1,48 @@
+
+	DateModified time.Time `bson:"date_modified" json:"date_modified"`
+	DateCreated  time.Time `bson:"date_created,omitempty" json:"date_created"`
 }
 
-// Address contains information about a user's address.
-type Address struct {
-	Type         int        `bson:"type" json:"type" validate:"required"`
-	LineOne      string     `bson:"line_one" json:"line_one" validate:"required"`
-	LineTwo      string     `bson:"line_two" json:"line_two,omitempty"`
-	City         string     `bson:"city" json:"city" validate:"required"`
-	State        string     `bson:"state" json:"state" validate:"required"`
-	Zipcode      string     `bson:"zipcode" json:"zipcode" validate:"required"`
-	Phone        string     `bson:"phone" json:"phone" validate:"required"`
-	DateModified *time.Time `bson:"date_modified" json:"date_modified"`
-	DateCreated  *time.Time `bson:"date_created,omitempty" json:"date_created"`
+// NewUser contains information needed to create a new User.
+type NewUser struct {
+	Name            string   `json:"name" validate:"required"`
+	Email           string   `json:"email" validate:"required"` // TODO(jlw) enforce uniqueness.
+	Roles           []string `json:"roles" validate:"required"` // TODO(jlw) ensure only admins can change other users' roles. Ensure only includes valid roles.
+	Password        string   `json:"password" validate:"required"`
+	PasswordConfirm string   `json:"password_confirm" validate:"eqfield=Password"`
 }
 
-// CreateUser contains information needed to create or update a user.
-type CreateUser struct {
-	UserType     int             `bson:"type" json:"type" validate:"required"`
-	FirstName    string          `bson:"first_name" json:"first_name" validate:"required"`
-	LastName     string          `bson:"last_name" json:"last_name" validate:"required"`
-	Email        string          `bson:"email" json:"email" validate:"required"`
-	Company      string          `bson:"company" json:"company" validate:"required"`
-	Addresses    []CreateAddress `bson:"addresses" json:"addresses" validate:"required"`
-	DateModified *time.Time      `bson:"date_modified" json:"-"`
+// UpdateUser defines what information may be provided to modify an existing
+// User. All fields are optional so clients can send just the fields they want
+// changed. It uses pointer fields so we can differentiate between a field that
+// was not provided and a field that was provided as explicitly blank. Normally
+// we do not want to use pointers to basic types but we make exceptions around
+// marshalling/unmarshalling.
+type UpdateUser struct {
+	Name            *string  `json:"name"`
+	Email           *string  `json:"email"` // TODO(jlw) enforce uniqueness.
+	Roles           []string `json:"roles"` // TODO(jlw) ensure only admins can change other users' roles. Ensure only includes valid roles.
+	Password        *string  `json:"password"`
+	PasswordConfirm *string  `json:"password_confirm" validate:"omitempty,eqfield=Password"`
```

Comment:
Is the password confirmation something that should be sent to the API? I have always seen this implemented on the frontend

Reply:
Yeah I have seen it at the API layer a lot. It isn't critical to the API though and could easily be delegated to the client I guess?

Reply:
Gotcha, either way